src/group_theory/iwasawa.lean

```diff
@@ -0,0 +1,486 @@
+/-
+Copyright (c) 2021 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as des
+cribed in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+-- import group_theory.subgroup.basic
+import group_theory.subgroup.pointwise
+import group_theory.coset
+import group_theory.quotient_group
+import group_theory.abelianization
+import group_theory.group_action.defs
+import group_theory.group_action.basic
+import group_theory.group_action.group
+import group_theory.group_action.conj_act
+
+/-!
+# Iwasawa criterion and simplicity of some groups
+
+Reference : https://www.math.ens.fr/~vanhaecke/tdalg2019/Correction_Iwasawa.pdf
+
+## Iwasawa criterion
+
+Let G be a group acting on a set X (`mul_action G X`).
+
+**Definition (class `is_transitive`)**. The action is *transitive*
+if there is x:X whose orbit is X
+
+**Definition (class `is_primitive`).** The action is *primitive*
+if it is transitive, and if all stabilizers G_x are maximal subgroups.
+(The class `subgroup.is_maximal` organizes the notion of maximal subgroups.)
+
+**Lemma (`is_primitive_of_two_trans`).**
+If the action is 2-transitive, then it is primitive.
+
+(TODO : define n-transitive ations (for n:ℕ) and proves the inductive results.)
+
+**Theorem (`Iwasawa_Criterion.Iwasawa_mk`, Iwasawa criterion).**
+Let G be a group with a primitive action on a set X.
+Assume given an Iwasawa structure, namely,
+for all x:X, an abelian subgroup T x of G such that
+1. ∀ g:G x:X, T (g x) = g (T x) g⁻¹
+2. The union of all T x generates G
+Then every normal subgroup of G that acts nontrivially on X contains
+the derived subgroup D(G) of G.
+
+These hypotheses are bundled in the structure `has_iwasawa_structure`.
+
+**Corollary (`Iwasawa_Criterion.is_simple`).**
+Let G be a perfect group (`commutator G = ⊤`)
+with a faithful and primitive action on a set X that admits an Iwasawa structure.
+Then G is simple.
+
+## Simplicity of special linear groups
+
+**Theorem (TODO).** Let K be a field and let n be an integer such that n ≥ 2.
+Unless n = 2 and K has ≤ 3 elements, the group PSL(n,K) is simple.
+
+
+## Simplicity of alternate groups
+
+**Theorem (TODO).** Let n be an integer such that n ≥ 5.
+The alternate group A(n) is simple.
+
+-/
+
+
+
+section Maximal_Subgroups
+variables {G : Type*} [group G]
+
+namespace subgroup
+/-- An subgroup is maximal if it is maximal in the collection of proper subgroups. -/
+class is_maximal (K : subgroup G) : Prop :=
```

Comment:
You make this a class, yet you never have `[is_maximal K]` anywhere.

Reply:
I just copied what was done for maximal ideals.

Reply:
I definitely agree with this being a class to be consistent - I imagine that instances will accumulate as people do more group theory.